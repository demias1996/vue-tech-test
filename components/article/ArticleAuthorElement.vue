<template>
    <div class="article-author-element d-flex align-items-center">
        <b-avatar class="mr-3" variant="primary">
            <b-icon-people-fill />
        </b-avatar>
        <span class="article-author-element__name">
            {{ authorName }}
        </span>
    </div>
</template>

<script>
    import { BIconPeopleFill } from 'bootstrap-vue';

    export default {
        name: 'article-author-element',

        components: { BIconPeopleFill },

        props: {
            authorName: {
                type: String,
                default: '',
            },
        },
    };
</script>

<style scoped>
    .article-author-element {
        color: #000000;
    }
</style>
