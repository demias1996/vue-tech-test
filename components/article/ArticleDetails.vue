<template>
    <section class="article-details">
        <p class="article-details__date">{{ articleDate }}</p>
        <h1 class="article-details__title">{{ article.title }}</h1>
        <article-author-element
            class="article-details__author"
            :author-name="article.author"
        />
        <b-img-lazy
            class="article-details__image"
            :src="article.image"
        />
        <p class="article-details__intro">{{ article.intro }}</p>
        <p class="article-details__text">{{ article.text }}</p>
    </section>
</template>

<script>
    import ArticleAuthorElement from '~/components/article/ArticleAuthorElement';

    export default {
        name: 'article-details',

        components: { ArticleAuthorElement },

        props: {
            article: {
                type: Object,
                required: true,
            },
        },

        computed: {
            /**
             * Returns properly formatted date of creation for article.
             *
             * @returns {string}
             */
            articleDate() {
                const date = new Date(this.article.createdAt);
                return date.toLocaleDateString();
            },
        },
    };
</script>

<style lang="scss" scoped>
    @import "~bootstrap/scss/functions";
    @import "~bootstrap/scss/variables";
    @import "~bootstrap/scss/mixins/_breakpoints";

    .article-details {
        padding-top: 15%;

        @include media-breakpoint-up(md) {
            padding: 5% 15%;
        }

        @include media-breakpoint-up(lg) {
            padding: 5% 25%;
        }

        &__date {
            text-align: center;
            color: #a9a9a9;
        }

        &__title,
        &__author,
        &__image {
            margin-bottom: 15px;
        }

        &__image {
            width: 100%;
            border-radius: 5px;
        }
    }
</style>
