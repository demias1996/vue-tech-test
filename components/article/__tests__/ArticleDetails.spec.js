import { shallowMount } from '@vue/test-utils';
import ArticleDetails from '@/components/article/ArticleDetails';
import { ArticleFixture } from '@/fixtures/ArticleFixture';
import ArticleAuthorElement from '@/components/article/ArticleAuthorElement'
import { BImgLazy } from 'bootstrap-vue'

/**
 * Returns wrapper for component shallowly mounted with required props.
 *
 * @param {object} propsData
 * @returns {Wrapper<Vue>}
 */
const shallowMountArticleDetails = (propsData = {}) =>
    shallowMount(ArticleDetails, { propsData: {
        article: { ...ArticleFixture },
        ...propsData,
    } });

describe('ArticleCard', () => {
    test('is a Vue instance', () => {
        const wrapper = shallowMountArticleDetails();
        expect(wrapper.vm).toBeTruthy();
    });

    test('is rendered correctly with required prop', () => {
        const wrapper = shallowMountArticleDetails();
        expect(wrapper.element).toMatchSnapshot();
    });

    test('is article date filled', () => {
        const wrapper = shallowMountArticleDetails();
        const articleDate = wrapper.find('.article-details__date');

        expect(articleDate.text()).toBe('9/15/2021');
    });

    test('is article title filled', () => {
        const wrapper = shallowMountArticleDetails();
        const titleElement = wrapper.find('.article-details__title');

        expect(titleElement.text()).toBe(ArticleFixture.title);
    });

    test('is article author name set', () => {
        const wrapper = shallowMountArticleDetails();
        const authorElement = wrapper.findComponent(ArticleAuthorElement);

        expect(authorElement.props().authorName).toBe(ArticleFixture.author);
    });

    test('is article image set', () => {
        const wrapper = shallowMountArticleDetails();
        const imageElement = wrapper.findComponent(BImgLazy);

        expect(imageElement.props().src).toBe(ArticleFixture.image);
    });

    test('is article intro filled', () => {
        const wrapper = shallowMountArticleDetails();
        const articleIntro = wrapper.find('.article-details__intro');

        expect(articleIntro.text()).toBe(ArticleFixture.intro);
    });

    test('is article text filled', () => {
        const wrapper = shallowMountArticleDetails();
        const articleText = wrapper.find('.article-details__text');

        expect(articleText.text()).toBe(ArticleFixture.text);
    });
});
