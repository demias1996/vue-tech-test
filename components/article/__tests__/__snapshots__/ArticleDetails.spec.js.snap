// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard is rendered correctly with required prop 1`] = `
<section
  class="article-details"
>
  <p
    class="article-details__date"
  >
    9/15/2021
  </p>
   
  <h1
    class="article-details__title"
  >
    Article title
  </h1>
   
  <article-author-element-stub
    authorname="Test author"
    class="article-details__author"
  />
   
  <b-img-lazy-stub
    blankcolor="transparent"
    class="article-details__image"
    offset="360"
    src="https://test-image-source"
  />
   
  <p
    class="article-details__intro"
  >
    Article intro
  </p>
   
  <p
    class="article-details__text"
  >
    Article text
  </p>
</section>
`;
